test/libsolidity/syntaxTests/nameAndTypeResolution/236_cyclic_binary_dependency.sol

```diff
@@ -2,4 +2,6 @@ contract A { function f() public { new B(); } }
 contract B { function f() public { new C(); } }
```

Comment:
Some more test ideas:
- Multiple creation cycles sharing one node (to test the case I mentioned in one of the comments where the error is reported only for that shared node and not separately for each contract).
- Creating the contract with `new` in its own constructor (I think we only have that for an ordinary function).
- A situation where two contracts just reference each other's names without doing anything (i.e. `f() public { C; }`). Currently this is allowed.
- Same but with `type(C)`.
- Two contracts calling each other without creating an actual dependency cycle (we might already have a test for this):
    ```solidity
    contract C { function foo(D _d) public { _d.foo(this); } }
    contract D { function foo(C _c) public { _c.foo(this); } }
    ```
- Same for libraries
    ```solidity
    library L1 { function foo() internal { L2.foo(); } }
    library L2 { function foo() internal { L1.foo(); } }
    ```
- Creation cycles involving both `new` and inheritance. For example:
    - `C is D`, `E is D`, `E` does `new C`
    - `C` does `new A`, `A` does `new D`, `D is C`.

Reply:
I am wondering if the libraries case _should_ work. In this new PR it is an error.

Reply:
* `C is D`, `E is D`, `E` does `new C` — This is currently no error. Should it be? 
* `C` does `new A`, `A` does `new D`, `D is C`. — This does produce an error.

Reply:
> I am wondering if the libraries case _should_ work. In this new PR it is an error.

I think it should because you do not need to compile one first to be able to compile the other. Both libraries have their own copies of both functions compiled into them.

> * `C is D`, `E is D`, `E` does `new C` — This is currently no error. Should it be? 

Interesting. Right, I think it's fine that it compiles because if we compile `C` first, `E` can use its bytecode just fine.

I'd still add it as a test case but the situation I had in mind is actually this: `C is D`, `E is D`, `C` does `new E`

> * `C` does `new A`, `A` does `new D`, `D is C`. — This does produce an error.

Great. This is what I would expect.